lib/api/normalize.js

```diff
@@ -0,0 +1,307 @@
+     */
+    resetAtomicBlocks(editorState, enabledTypes) {
+        const content = editorState.getCurrentContent();
+        const blockMap = content.getBlockMap();
+        let blocks = blockMap;
+
+        const normalisedBlocks = blocks
+            .filter(
+                block =>
+                    block.getType() === BLOCK_TYPE.ATOMIC &&
+                    (block.getText() !== ' ' ||
+                        block.getInlineStyleAt(0).size !== 0),
+            )
+            .map(block => {
+                // Retain only the first character, and remove all of its styles.
+                const chars = block
+                    .getCharacterList()
+                    .slice(0, 1)
+                    .map(char => {
+                        let newChar = char;
+
+                        char.getStyle().forEach(type => {
+                            newChar = CharacterMetadata.removeStyle(
+                                newChar,
+                                type,
+                            );
+                        });
+
+                        return newChar;
+                    });
+                return block.merge({
+                    text: ' ',
+                    characterList: chars,
+                });
+            });
+
+        if (normalisedBlocks.size !== 0) {
+            blocks = blockMap.merge(normalisedBlocks);
+        }
+
+        const resetBlocks = blocks
+            .filter(block => block.getType() === BLOCK_TYPE.ATOMIC)
+            .filter(block => {
+                const entityKey = block.getEntityAt(0);
+                let shouldReset = false;
+
+                if (entityKey) {
+                    const entityType = content.getEntity(entityKey).getType();
+
+                    shouldReset = !enabledTypes.includes(entityType);
+                }
+
+                return shouldReset;
+            })
+            .map(block => block.set('type', BLOCK_TYPE.UNSTYLED));
+
+        if (resetBlocks.size !== 0) {
+            blocks = blockMap.merge(resetBlocks);
+        }
+
+        return EditorState.set(editorState, {
+            currentContent: content.merge({
+                blockMap: blockMap.merge(blocks),
+            }),
+        });
+    },
+
+    /**
+     * Reset all entity types (images, links, documents, embeds) that are unavailable.
+     */
+    filterEntityType(editorState, enabledTypes) {
+        const content = editorState.getCurrentContent();
+        const blockMap = content.getBlockMap();
+
+        /**
+         * Removes entities from the character list if the character entity isn't enabled.
+         * Also removes image entities placed outside of atomic blocks, which can happen
+         * on paste.
+         * A better approach would probably be to split the block where the image is and
+         * create an atomic block there, but that's another story. This is what Draft.js
+         * does when the copy-paste is all within one editor.
```

Comment:
This may not be that complicated with the `Modifier` method that does this, but another time.